external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -250,13 +432,17 @@ impl Compatibility {
                 )
             {
                 match old_func.visibility {
-                    Visibility::Friend => friend_linking = false,
-                    Visibility::Public => datatype_and_function_linking = false,
+                    Visibility::Friend => {
+                        context.function_previously_friend(old_func);
+                    }
+                    Visibility::Public => {
+                        context.function_visibility_mismatch(old_func, new_func);
+                    }
                     Visibility::Private => (),
                 }
 
                 if old_func.is_entry {
-                    entry_linking = false;
+                    context.function_entry_compatibility(old_func, new_func);
                 }
             }
```

Comment:
These will need new functions added to the context. They are not about the function's visibility changing, but about the function's signature changing in various contexts where its signature may be depended upon.

Reply:
seems like the if statement could be broken down too?

Reply:
The problem I'm seeing with breaking down is we are creating
2 visibility types(friend,public) * 3 signature checks (params, return, types) = 6 new functions
this case might need a little logic inside so we don't add a ton of functions to mark signature mismatching.

I see 3 options:
- keep it simple just error with "signature mismatch" use `_friend`, and `_public` functions so the proper flags are set 
- create all the functions for each visibility and signature issue (6 functions) highest fidelity but more boiler plate
- use visibility enum as an argument and match inside ExecutionCompatibilityMode functions `sig_mismatch_type`, `sig_mismatch_return`, `sig_mismatch_params` and have 3 functions, let the function keep all the information but don't create so many functions to capture all the possibilities.

Reply:
I like that last option best!

Reply:
In fact, I think we can go further, and say this is generally a `signature_mismatch(old_func, new_func)`. `ExecutionCompatibilityMode` would then include logic to figure out which kind of linkage this applies to based on the old function's visibility.